test/compiler/inline.jl

```diff
@@ -319,6 +319,62 @@ function f37555(x::Int; kwargs...)
 end
 @test f37555(1) == 1
 
+# Issue #18773 - Specifying `@noinline` at callsite
+# Ensure `@noinline` in caller overrides `@inline` in callee
+@inline f18773(x) = x
+g18773(x) = @noinline f18773(x)
+let ci = code_typed(g18773, Tuple{Int})[1].first
+    @test length(ci.code) == 2 &&
+        isexpr(ci.code[1], :invoke) &&
+        ci.code[1].args[1].def.name == :f18773
+end
+# Test that `@noinline` works across entire expression
+h18773(x) = @noinline f18773(x) + f18773(x)
+let ci = code_typed(h18773, Tuple{Int})[1].first
+    @test length(ci.code) == 4 &&
+        isexpr(ci.code[1], :invoke) &&
+        ci.code[1].args[1].def.name == :f18773 &&
+        isexpr(ci.code[2], :invoke) &&
+        ci.code[2].args[1].def.name == :f18773
+end
+
+# Test inlining/noinlining of code within `do` blocks
+@inline simple_caller(a) = a()
+function do_inline(x)
+    simple_caller() do
+        @inline
+        # Tests `@inline`'s ability to override the lack of inline
+        # that these `println` statements would have caused
+        println(stdout, "Hello")
+        println(stdout, "World")
+        println(stdout, "Hello")
+        println(stdout, "World")
+        println(stdout, "Hello")
+        println(stdout, "World")
+        println(stdout, "Hello")
+        println(stdout, "World")
+        println(stdout, "Hello")
+        println(stdout, "World")
+        println(stdout, "Hello")
+        println(stdout, "World")
+        x
+    end
+end
+let ci = code_typed(do_inline, Tuple{Int})[1].first
+    # A long body indicates inlining occurred
+    @test length(ci.code) == 25
+end
```

Comment:
```suggestion
@inline simple_caller(a) = a()
function do_inline(x)
    simple_caller() do
        @inline
        # this call won't be resolved and thus will prevent inlining to happen if we don't
        # annotate `@inline` at the top of this anonymous function body
        return unresolved_call(x)
    end
end
let ci = first(@code_typed do_inline(1))
    # what we test here is that both `simple_caller` and the anonymous function that the
    # `do` block creates are inlined away, and as a result there is only the unresolved call
    @test all(ci.code) do x
        !(isexpr(x, :invoke) && x.args[1].def.name === :simple_caller) &&
        !(isexpr(x, :invoke) && startswith(string(x.args[1].def.name), '#'))
    end
end
```

This would be more minimal and robust.